src/abe_policy/dimension.rs

```diff
@@ -123,33 +138,49 @@ impl Dimension {
     ///
     /// * `dim` - The `DimensionBuilder` to base the dimension on.
     /// * `seed_id` - A mutable reference to a seed ID used for generating
-    ///   unique IDs for attributes.
+    ///   unique values for attributes.
     pub fn new(dim: &DimensionBuilder, seed_id: &mut u32) -> Self {
-        let attributes_mapping = dim
-            .attributes_properties
-            .iter()
-            .map(|attr| {
-                (
-                    attr.name.clone(),
-                    AttributeParameters::new(attr.encryption_hint, seed_id),
-                )
-            })
-            .collect();
+        let attributes_mapping = dim.attributes_properties.iter().map(|attr| {
+            (
+                attr.name.clone(),
+                AttributeParameters::new(attr.encryption_hint, seed_id),
+            )
+        });
 
         match dim.hierarchical {
-            true => Self {
-                order: Some(
-                    dim.attributes_properties
-                        .iter()
-                        .map(|attr| attr.name.clone())
-                        .collect(),
-                ),
-                attributes: attributes_mapping,
-            },
-            false => Self {
-                order: None,
-                attributes: attributes_mapping,
-            },
+            true => Self::Ordered(attributes_mapping.collect()),
+            false => Self::Unordered(attributes_mapping.collect()),
+        }
+    }
+
+    pub fn nb_attributes(&self) -> usize {
+        match self {
+            Self::Unordered(attributes) => attributes.len(),
+            Self::Ordered(attributes) => attributes.len(),
+        }
+    }
+
+    pub fn is_ordered(&self) -> bool {
+        match self {
+            Self::Unordered(_) => false,
+            Self::Ordered(_) => true,
+        }
+    }
+
+    /// Returns an iterator over the attributes name.
+    /// If the dimension is ordered, the names are returned in this order,
+    /// otherwise they are returned in arbitrary order.
+    pub fn get_attributes_name(&self) -> Box<dyn '_ + Iterator<Item = &AttributeName>> {
+        match self {
+            Self::Unordered(attributes) => Box::new(attributes.keys()),
+            Self::Ordered(attributes) => Box::new(attributes.keys()),
```

Comment:
note that if `attributes` is the same type on each arm you can write:
```suggestion
            Self::Unordered(attributes) | Self::Ordered(attributes) => Box::new(attributes.keys()),
```